### PR TITLE
[RISCV][GlobalISel] Select G_PTR_ADD

### PR DESCRIPTION
**Pull request by nitinjohnraj**
We lower G_PTR_ADD to G_ADD before selection. We choose to do this here and not post-legalization because G_PTR_ADD can still be combined with other instructions during selection.

### CODE REVIEW

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 1]
```diff
@@ -45,6 +45,16 @@ class RISCVInstructionSelector : public InstructionSelector {
   getRegClassForTypeOnBank(LLT Ty, const RegisterBank &RB) const;
 
   bool selectImpl(MachineInstr &I, CodeGenCoverage &CoverageInfo) const;
+
+  // A lowering phase that runs before any selection attempts.
+  // Returns true if the instruction was modified.
+  bool preISelLower(MachineInstr &MI, MachineIRBuilder &MIB,
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:51]: I don't think we need the return value here, as opposed to its AArch64 counterpart

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 2]
```diff
@@ -225,9 +237,13 @@ bool RISCVInstructionSelector::select(MachineInstr &MI) {
 
   switch (Opc) {
   case TargetOpcode::G_ANYEXT:
-  case TargetOpcode::G_TRUNC:
+  case TargetOpcode::G_PTRTOINT:
+  case TargetOpcode::G_TRUNC: {
     MI.setDesc(TII.get(TargetOpcode::COPY));
+    bool Selected = selectCopy(MI, MRI);
+    assert(Selected && "Could not select copy!");
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:244]: compilers might complain about unused `Selected` variable when assertion is disabled. So maybe we can `(void) Selected` or `return Selected;`.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 3]
```diff
@@ -245,6 +261,39 @@ bool RISCVInstructionSelector::select(MachineInstr &MI) {
   }
 }
 
+bool RISCVInstructionSelector::replacePtrWithInt(MachineOperand &Op,
+                                                 MachineIRBuilder &MIB,
+                                                 MachineRegisterInfo &MRI) {
+  assert(Op.isReg() && "Operand is not a register!");
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:267]: `getReg` will do the assertion of `isReg` so I don't think this is needed here.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp (hunk 2 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:244]: This piece of code has been rewritten recently. Please rebase.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 4]
```diff
@@ -130,12 +140,14 @@ bool RISCVInstructionSelector::earlySelectShift(
 }
 
 bool RISCVInstructionSelector::select(MachineInstr &MI) {
-  unsigned Opc = MI.getOpcode();
   MachineBasicBlock &MBB = *MI.getParent();
   MachineFunction &MF = *MBB.getParent();
   MachineRegisterInfo &MRI = MF.getRegInfo();
   MachineIRBuilder MIB(MI);
 
+  preISelLower(MI, MIB, MRI);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:148]: preISelLower returns a bool. Do we need return false if that bool is false?

**Comment by topperc** [comment]
Can you rebase this and get rid of the merge commit. Its angering the github display. I'm seeing 100 commits

**Comment by ldionne** [comment]
This also caused the wrong teams to be requested a review from.

**Comment by nitinjohnraj** [comment]
Got it, I'll be more careful of that in the future

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**